File: src/main/java/catchla/yep/view/HeaderSpaceLayout.java
/*
 * Copyright (c) 2015. Catch Inc,
 */

package catchla.yep.view;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.FrameLayout;

/**
 * Created by mariotaku on 15/5/5.
 */
public class HeaderSpaceLayout extends FrameLayout {

    private int mMinusTop;
    private float mWidthHeightRatio = 1;

    public HeaderSpaceLayout(Context context) {
        super(context);
    }

    public HeaderSpaceLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public HeaderSpaceLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public void setWidthHeightRatio(float ratio) {
        mWidthHeightRatio = ratio;
        requestLayout();
    }

    public void setMinusTop(int minusTop) {
        mMinusTop = minusTop;
    }

    @Override
    protected void onMeasure(final int widthMeasureSpec, final int heightMeasureSpec) {
        final int width = MeasureSpec.getSize(widthMeasureSpec);
        final int height = Math.round(width * mWidthHeightRatio) - mMinusTop;
        final int hSpec = MeasureSpec.makeMeasureSpec(height, MeasureSpec.getMode(heightMeasureSpec));
        super.onMeasure(widthMeasureSpec, hSpec);
        setMeasuredDimension(width, height);
    }

}

File: src/main/kotlin/catchla/yep/view/HeaderSpaceLayout.kt
/*
 * Copyright (c) 2015. Catch Inc,
 */

package catchla.yep.view

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout

/**
 * Created by mariotaku on 15/5/5.
 */
class HeaderSpaceLayout : FrameLayout {

    var minusTop: Int = 0
    var widthHeightRatio = 1f
        set(ratio) {
            field = ratio
            requestLayout()
        }

    constructor(context: Context) : super(context) {
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val width = View.MeasureSpec.getSize(widthMeasureSpec)
        val height = Math.round(width * this.widthHeightRatio) - minusTop
        val hSpec = View.MeasureSpec.makeMeasureSpec(height, View.MeasureSpec.getMode(heightMeasureSpec))
        super.onMeasure(widthMeasureSpec, hSpec)
        setMeasuredDimension(width, height)
    }

}
